################################################################################
SHORT DESCRIPTION: 
################################################################################
The v4v firewall patch.

################################################################################
LONG DESCRIPTION: 
################################################################################
The implemenation of the v4v firewall/viptables. See the OpenXT Wiki for more
information on v4v.

################################################################################
CHANGELOG 
################################################################################

################################################################################
REMOVAL 
################################################################################

################################################################################
UPSTREAM PLAN
################################################################################
Yes if we can.

################################################################################
INTERNAL DEPENDENCIES 
################################################################################

################################################################################
PATCHES 
################################################################################
--- a/xen/common/v4v.c
+++ b/xen/common/v4v.c
@@ -68,6 +68,7 @@ static DEFINE_RWLOCK (v4v_lock); /* L1 *
 /* guest ring_data and the pending list. To take L3 you must */
 /* already have R(L2). W(L2) implies L3 */
 
+struct list_head viprules = LIST_HEAD_INIT(viprules);
 
 /*Debugs*/
 
@@ -1384,6 +1385,216 @@ v4v_ring_add (struct domain *d, XEN_GUES
   return ret;
 }
 
+/**************************** VIPTables ***************************/
+
+static void
+v4v_viptables_print_rule (struct v4v_viptables_rule_node *rule)
+{
+  if (rule == NULL)
+    {
+      printk("(null)\n");
+      return;
+    }
+
+  if (rule->accept == 1)
+    printk("ACCEPT");
+  else
+    printk("REJECT");
+
+  printk(" ");
+
+  if (rule->src.domain == DOMID_INVALID)
+    printk("*");
+  else
+    printk("%i", rule->src.domain);
+
+  printk(":");
+
+  if (rule->src.port == -1)
+    printk("*");
+  else
+    printk("%u", rule->src.port);
+
+  printk(" -> ");
+
+  if (rule->dst.domain == DOMID_INVALID)
+    printk("*");
+  else
+    printk("%i", rule->dst.domain);
+
+  printk(":");
+
+  if (rule->dst.port == -1)
+    printk("*");
+  else
+    printk("%u", rule->dst.port);
+
+  printk("\n");
+}
+
+static int
+v4v_viptables_add (struct domain *src_d, XEN_GUEST_HANDLE(v4v_viptables_rule_t) rule,
+                   int32_t position)
+{
+  struct v4v_viptables_rule_node* new;
+  struct list_head* tmp;
+
+  /* First rule is n.1 */
+  position--;
+
+  new = v4v_xmalloc (struct v4v_viptables_rule_node);
+
+  if (copy_field_from_guest (new, rule, src))
+    return -EFAULT;
+  if (copy_field_from_guest (new, rule, dst))
+    return -EFAULT;
+  if (copy_field_from_guest (new, rule, accept))
+    return -EFAULT;
+
+  printk(KERN_ERR "VIPTables: ");
+  v4v_viptables_print_rule(new);
+
+  tmp = &viprules;
+  while (position != 0 && tmp->next != &viprules)
+    {
+      tmp = tmp->next;
+      position--;
+    }
+  list_add(&new->list, tmp);
+
+  return 0;
+}
+
+static int
+v4v_viptables_del (struct domain *src_d, XEN_GUEST_HANDLE(v4v_viptables_rule_t) rule,
+                   int32_t position)
+{
+  struct list_head *tmp = NULL;
+  struct list_head *next = NULL;
+  struct v4v_viptables_rule_node *node;
+  struct v4v_viptables_rule *r;
+
+  if (position != -1)
+    {
+      /* We want to delete the rule number <position> */
+      tmp = &viprules;
+      while (position != 0 && tmp->next != &viprules)
+        {
+          tmp = tmp->next;
+          position--;
+        }
+    }
+  else if (!guest_handle_is_null(rule))
+    {
+      /* We want to delete the rule <rule> */
+      r = v4v_xmalloc (struct v4v_viptables_rule);
+
+      if (copy_field_from_guest (r, rule, src))
+        return -EFAULT;
+      if (copy_field_from_guest (r, rule, dst))
+        return -EFAULT;
+      if (copy_field_from_guest (r, rule, accept))
+        return -EFAULT;
+
+      list_for_each(tmp, &viprules)
+        {
+          node = list_entry(tmp, struct v4v_viptables_rule_node, list);
+
+          if ((node->src.domain == r->src.domain) &&
+              (node->src.port   == r->src.port)   &&
+              (node->dst.domain == r->dst.domain) &&
+              (node->dst.port   == r->dst.port))
+            {
+              position = 0;
+              break;
+            }
+        }
+      v4v_xfree(r);
+    }
+  else
+    {
+      /* We want to flush the rules! */
+      printk(KERN_ERR "VIPTables: flushing rules\n");
+      list_for_each_safe(tmp, next, &viprules)
+        {
+          node = list_entry(tmp, struct v4v_viptables_rule_node, list);
+          list_del(tmp);
+          v4v_xfree(node);
+        }
+    }
+
+  if (position == 0 && tmp != &viprules)
+    {
+      printk(KERN_ERR "VIPTables: deleting rule: ");
+      node = list_entry(tmp, struct v4v_viptables_rule_node, list);
+      v4v_viptables_print_rule(node);
+      list_del(tmp);
+      v4v_xfree(node);
+    }
+
+  return 0;
+}
+
+static int
+v4v_viptables_list (struct domain *src_d, XEN_GUEST_HANDLE(v4v_viptables_list_t) list_hnd)
+{
+  struct list_head *ptr;
+  struct v4v_viptables_rule_node *node;
+  struct v4v_viptables_list rules_list;
+
+  memset(&rules_list, 0, sizeof (rules_list));
+  if (copy_field_from_guest (&rules_list, list_hnd, nb_rules))
+      return -EFAULT;
+
+  ptr = viprules.next;
+  while (rules_list.nb_rules != 0 && ptr != &viprules)
+  {
+      ptr = ptr->next;
+      rules_list.nb_rules--;
+  }
+
+  if (rules_list.nb_rules != 0)
+      return -EFAULT;
+
+  while (rules_list.nb_rules < V4V_VIPTABLES_LIST_SIZE &&
+         ptr != &viprules)
+  {
+      node = list_entry(ptr, struct v4v_viptables_rule_node, list);
+
+      rules_list.rules[rules_list.nb_rules].src = node->src;
+      rules_list.rules[rules_list.nb_rules].dst = node->dst;
+      rules_list.rules[rules_list.nb_rules].accept = node->accept;
+
+      rules_list.nb_rules++;
+      ptr = ptr->next;
+  }
+
+  if (copy_to_guest(list_hnd, &rules_list, 1))
+      return -EFAULT;
+
+  return 0;
+}
+
+static size_t
+v4v_viptables_check (v4v_addr_t * src, v4v_addr_t * dst)
+{
+  struct list_head *ptr;
+  struct v4v_viptables_rule_node *node;
+
+  list_for_each(ptr, &viprules)
+    {
+      node = list_entry(ptr, struct v4v_viptables_rule_node, list);
+
+      if ((node->src.domain == DOMID_INVALID || node->src.domain == src->domain) &&
+          (node->src.port   == -1            || node->src.port   == src->port)   &&
+          (node->dst.domain == DOMID_INVALID || node->dst.domain == dst->domain) &&
+          (node->dst.port   == -1            || node->dst.port   == dst->port))
+        return !node->accept;
+    }
+
+  /* Defaulting to ACCEPT */
+  return 0;
+}
 
 /**************************** io ***************************/
 
@@ -1543,6 +1754,15 @@ v4v_send (struct domain *src_d, v4v_addr
       ret = -EPERM;
       goto out;
     }
+  /* VIPTables*/
+  if (v4v_viptables_check(src_addr, dst_addr) != 0)
+    {
+      printk(KERN_ERR "V4V: VIPTables REJECTED %i:%u -> %i:%u\n",
+             src_addr->domain, src_addr->port,
+             dst_addr->domain, dst_addr->port);
+      ret = -ENOTCONN;
+      goto out;
+    }
 
   do
     {
@@ -1645,6 +1865,15 @@ v4v_sendv (struct domain *src_d, v4v_add
       ret = -ENOTCONN;
       goto out;
     }
+  /* VIPTables*/
+  if (v4v_viptables_check(src_addr, dst_addr) != 0)
+    {
+      printk(KERN_ERR "V4V: VIPTables REJECTED %i:%u -> %i:%u\n",
+             src_addr->domain, src_addr->port,
+             dst_addr->domain, dst_addr->port);
+      ret = -EPERM;
+      goto out;
+    }
 
   do
     {
@@ -1802,6 +2031,38 @@ do_v4v_op (int cmd, XEN_GUEST_HANDLE (vo
         rc = v4v_notify (d, ring_data_hnd);
         break;
       }
+    case V4VOP_viptables_add:
+      {
+        uint32_t position = arg4;
+        XEN_GUEST_HANDLE (v4v_viptables_rule_t) rule_hnd =
+          guest_handle_cast (arg1, v4v_viptables_rule_t);
+        rc = -EPERM;
+        if (!d->is_privileged)
+            goto out;
+        rc = v4v_viptables_add (d, rule_hnd, position);
+        break;
+      }
+    case V4VOP_viptables_del:
+      {
+        uint32_t position = arg4;
+        XEN_GUEST_HANDLE (v4v_viptables_rule_t) rule_hnd =
+          guest_handle_cast (arg1, v4v_viptables_rule_t);
+        rc = -EPERM;
+        if (!d->is_privileged)
+            goto out;
+        rc = v4v_viptables_del (d, rule_hnd, position);
+        break;
+      }
+    case V4VOP_viptables_list:
+      {
+        XEN_GUEST_HANDLE (v4v_viptables_list_t) rules_list_hnd =
+            guest_handle_cast(arg1, v4v_viptables_list_t);
+        rc = -EPERM;
+        if (!d->is_privileged)
+            goto out;
+        rc = v4v_viptables_list (d, rules_list_hnd);
+        break;
+      }
     default:
       rc = -ENOSYS;
       break;
--- a/xen/include/public/v4v.h
+++ b/xen/include/public/v4v.h
@@ -94,6 +94,25 @@ typedef struct v4v_addr
 
 DEFINE_XEN_GUEST_HANDLE (v4v_addr_t);
 
+typedef struct v4v_viptables_rule
+{
+    struct v4v_addr src;
+    struct v4v_addr dst;
+    uint32_t accept;
+} V4V_PACKED v4v_viptables_rule_t;
+
+DEFINE_XEN_GUEST_HANDLE (v4v_viptables_rule_t);
+
+#define V4V_VIPTABLES_LIST_SIZE 8
+
+typedef struct v4v_viptables_list
+{
+    struct v4v_viptables_rule rules[V4V_VIPTABLES_LIST_SIZE];
+    uint32_t nb_rules;
+} V4V_PACKED v4v_viptables_list_t;
+
+DEFINE_XEN_GUEST_HANDLE (v4v_viptables_list_t);
+
 typedef struct v4v_ring_id
 {
     struct v4v_addr addr;
@@ -250,6 +269,9 @@ struct v4v_ring_message_header
 /* Identical to V4VOP_send except rather than buf and len it takes
  * an array of v4v_iov_t and a length of the array */
 
+#define V4VOP_viptables_add     6
+#define V4VOP_viptables_del     7
+#define V4VOP_viptables_list    8
 
 #if !defined(__GNUC__)
 #pragma warning(pop)
--- a/xen/include/xen/v4v.h
+++ b/xen/include/xen/v4v.h
@@ -36,6 +36,13 @@ struct v4v_pending_ent
 };
 
 
+struct v4v_viptables_rule_node
+{
+    struct list_head list;
+    struct v4v_addr src;
+    struct v4v_addr dst;
+    uint32_t accept;
+};
 
 struct v4v_ring_info
 {
